src/services/services.ts

```diff
@@ -1440,9 +1378,15 @@ namespace ts {
             };
             program = createProgram(options);
 
-            // hostCache is captured in the closure for 'getOrCreateSourceFile' but it should not be used past this point.
-            // It needs to be cleared to allow all collected snapshots to be released
-            hostCache = undefined;
+            // 'getOrCreateSourceFile' depends on caching but should be used past this point.
+            // After this point, the cache needs to be cleared to allow all collected snapshots to be released
+            cachingInEffect = false;
+            compilerHost.getSourceFile = originalGetSourceFile;
+            compilerHost.readFile = originalReadFile;
+            compilerHost.fileExists = originalFileExists;
+            compilerHost.directoryExists = originalDirectoryExists;
+            compilerHost.createDirectory = originalCreateDirectory;
+            compilerHost.writeFile = originalWriteFile!;

```

Comment:
Instead of having flag to see if caching is in effect, can we set compilerHost to undefined here and check that instead of flag to see if caching is in effect.
```suggestion
            compilerHost = undefined;
```

Reply:
Oops, missed this comment.  I'm happy to give that a shot, but I found it pretty difficult to track the lifetime of `compilerHost` so I'm basically just taking your word for it.

Reply:
Done, though I personally prefer the version without the extra bangs.